lib/puppet/pops/types/types.rb

```diff
@@ -603,6 +671,111 @@ def _assignable?(o)
       end
     end
 
+    class PIterableType < PAnyType
+      attr_reader :element_type
+
+      def initialize(type)
+        @element_type = type
+      end
+
+      def instance?(o)
+        if @element_type.nil? || @element_type.assignable?(PAnyType::DEFAULT)
+          # Any element_type will do
+          case o
+          when Iterable, String, Hash, Enumerable, Enumerator, Range, PEnumType
```

Comment:
why are Enumerable and Enumerator here? Shouldn't this be Array? We are not really supporting any Ruby type that can be enumerated since we do not know what they return. Although we have the same problem with Hash and Array we have chosen to use the Ruby objects as Puppet runtime objects, so we have to live with those.